corehq/apps/hqadmin/reports.py

```diff
@@ -333,3 +338,174 @@ def _shorten_and_hyperlink_commit(self, commit_sha):
+        return [
+            domain,
+            case_count,
+            form_count,
+            self._restrict_rebuild_column_data(domain, case_count, form_count),
+        ]
+
+    def _should_restrict_rebuild(self, case_count, form_count):
+        return case_count >= self.UCR_RESTRICTION_THRESHOLD or form_count >= self.UCR_RESTRICTION_THRESHOLD
+
+    @staticmethod
+    @memoized
+    def _rebuild_restricted_ff_enabled(domain):
+        return RESTRICT_DATA_SOURCE_REBUILD.enabled(domain)
+
+    @property
+    def _show_ff_enabled_domains(self):
+        return self.restriction_ff_status == UCRRestrictionFFStatus.Enabled.name
+
+    @property
+    def _show_ff_disabled_domains(self):
+        return self.restriction_ff_status == UCRRestrictionFFStatus.NotEnabled.name
+
+    @property
+    def _show_should_enable_ff_domains(self):
+        return self.restriction_ff_status == UCRRestrictionFFStatus.ShouldEnable.name
+
+    @property
+    def _show_should_disable_ff_domains(self):
+        return self.restriction_ff_status == UCRRestrictionFFStatus.CanDisable.name
+
+    @property
+    def _show_all_domains(self):
+        return not self.restriction_ff_status
+
+    def _restrict_rebuild_column_data(self, domain, case_count, form_count):
+        from django.utils.safestring import mark_safe
+        from corehq.apps.toggle_ui.views import ToggleEditView
+
+        restriction_ff_enabled = self._rebuild_restricted_ff_enabled(domain)
+        toggle_edit_url = reverse(ToggleEditView.urlname, args=(RESTRICT_DATA_SOURCE_REBUILD.slug,))
+
+        if self._should_restrict_rebuild(case_count, form_count):
+            if not restriction_ff_enabled:
+                return mark_safe(f"""
+                    <a href={toggle_edit_url}>{gettext_lazy("Rebuild restriction required")}</a>
+                """)
+            return gettext_lazy("Rebuild restricted")
+
+        if restriction_ff_enabled:
+            return mark_safe(f"""
+                <a href={toggle_edit_url}>{gettext_lazy("Rebuild restriction not required")}</a>
+            """)
+        return gettext_lazy("No rebuild restriction required")
+
+
+class UCRDataLoadReport(AdminReport):
+    slug = 'ucr_data_load'
+    name = gettext_lazy("UCR Domains Data Report")
+
+    fields = [
+        'corehq.apps.reports.filters.select.UCRRebuildStatusFilter',
+    ]
+    emailable = False
+    exportable = False
+    default_rows = 10
+    ajax_pagination = True
+
+    def __init__(self, request, *args, **kwargs):
+        self.table_data = UCRRebuildRestrictionTable(
+            restriction_ff_status=request.GET.get('ucr_rebuild_restriction')
+        )
+        super().__init__(request, *args, **kwargs)
+
+    @property
+    def headers(self):
+        return self.table_data.headers
+
+    @property
+    def rows(self):
+        start = self.pagination.start
+        end = self.pagination.start + self.pagination.count
+        return self.table_data.rows[start:end]
```

Comment:
does this mean that we are creating all rows on each page and then picking the ones for the page?
Don't think we should be doing that.

Reply:
Honestly, I don't think we need pagination on this report since we are only fetching domains with UCR feature flag enabled which should not be many.

Reply:
@mkangia 

> does this mean that we are creating all rows on each page and then picking the ones for the page?

Correct. I do feel slightly uncomfortable about that, but I'm not sure how else I could achieve pagination, considering that I need to pull the domains first and then check if the domain applies based on the filter value. In this sense I need to construct all the applicable rows, but then paginate for the UI.

Note: I did check with an AE on whether the filters are necessary in the first place (since they seem kinda nice-to-have) but it was confirmed that they would be very helpful.

Reply:
> I don't think we need pagination on this report since we are only fetching domains with UCR feature flag enabled which should not be many.

The domains on prod [having this FF](https://www.commcarehq.org/hq/flags/edit/user_reports/) is already quite long.

Reply:
Just to be clear on what's happening here:
1. A list of domains are fetched which has the UCR feature flag enabled
2. The domains are iterated and a subset of domains are selected based on the filters (this is something that needs to happen on the application level as there's various things to check)
3. This reduced (might be little, might be a lot) list of domains are then only paginated.

Considering this, if we get rid of pagination there won't be any performance increase except the time it takes to slice a list in python, as pagination can only happens as a last step for this feature (not ideal, but don't see a way around it).

If we get rid of the filters the UX is deteriorated significantly. 

The trade-off is thus the performance on HQ vs time spent by AE's scrolling down a list trying to determine if some domains need a feature flag enabled.

I'm leaning towards trading off performance on HQ, as its not a significant enough concern in my opinion. To give an idea of why I'm thinking this way: check out performance [on staging](https://staging.commcarehq.org/hq/admin/ucr_data_load/) for this report. Production has just a little more than double the amount of records.

Reply:
one small correction here / correct me if I am wrong:
> The domains are iterated and a subset of domains are selected based on the filters (this is something that needs to happen on the application level as there's various things to check)

The domains are iterated and data (case and form count) is fetched for all of them one at a time.
For each domain, It is after the data fetch, we decide if we want to list the domain or not based on filters.
And then once we have the final list of valid domains based on filters, we then paginate.

I agree that performance implication isn't noticeable on staging and then possibly would be same on production.
But I would not consider that as a validation to go with this approach.

If you look at the code, it is like, if these were SQL records, you are fetching all records from DB and then paginating using a list, instead of asking SQL to give you a set of records. And its reasonable because the table is small?
So, I believe, irrespective of the performance implication, you wont go with the above approach to paginate the list instead of via SQL.

The program is fetching and processing all domains for each page anyway, so if we just skip the pagination, the page itself won't become any fast for any single page, but the user won't do multiple requests (different pages) anymore.
So, if you wish to keep the filters, we should skip pagination. Initially, we noticed that the number of UCRs is high, but then the program is going through all of them anyway even now.

Reply:
> The domains are iterated and data (case and form count) is fetched for all of them one at a time.

This is not true anymore. It was initially the way I did it by changed it [here](https://github.com/dimagi/commcare-hq/pull/34813/commits/76f9c74842c0f8ddabeb2c4bc766acb05a5bffb5). Now all the domains' case data is fetched in one query to ES and the forms data is fetched in another query to ES.

 > The program is fetching and processing all domains for each page anyway, so if we just skip the pagination, the page itself won't become any fast for any single page, but the user won't do multiple requests (different pages) anymore.

This is true. I see your point. Yeah, so if we skip pagination and keep the filters I can now see how that would be an improvement to the current approach. I'll update

Reply:
Reverted pagination [here](https://github.com/dimagi/commcare-hq/pull/34813/commits/7860d98bdcd80f51798849bd110a2ae94334ffdb).

Reply:
I noticed that the page still looks paginated (maybe this is the standard look) but all records are loaded in the table, since there's no data loading going on when I switch between pages.

Reply:
Do you need to set `disable_pagination=True` on the class?